### ISSUE TITLE
Rule: disallow "[", "(" at the start of the line

### ISSUE THREAD

**Issue by rlidwka** [eslint/eslint#746]
This rule should require a semicolon in front of a leading `(` or `[` at the start of the line. This prevents the expression from being interpreted as a function call or property access, respectively. ([source](https://github.com/npm/npm/blob/master/doc/misc/npm-coding-style.md#semicolons))

```js
// these cases should be flagged:
(x || y).doSomething()
[a, b, c].forEach(doSomething)

// these shouldn't:
;(x || y).doSomething()
;[a, b, c].forEach(doSomething)
```

Makes sense only if "semi" rule is set to "never", so it might be built-in to semi rule instead of making it a new one.

Related: https://github.com/eslint/eslint/issues/618#issuecomment-35455986.

**Comment by mattdesl** [comment]
IMHO this should be added as a separate rule, disabled by default but strongly recommended to those choosing to use `"semi": [2, "never"]`. 

Without `semi` and `no-spaced-func` the following doesn't report any lint errors, even though it is very broken.

```js
var foo = [2, 3, [4, 5]]
(Array.isArray(foo) ? foo : [foo])
  .filter(Boolean)
  .forEach(function(arg){
    console.log(arg)
  })
```

**Comment by nzakas** [comment]
I'm not sure I can buy this as a completely separate rule. Would it also flag this?

```js
(function() {

}());
```

**Comment by mattdesl** [comment]
IMHO yes it should since it can fall into the same trap. Even in ASI codebases (like npm, VueJS, etc) it can be useful to have a rule like this to discourage contributors from adopting a style you wish to discourage. It would (a) flag the above code as erroneous and (b) force contributors to rewrite it like so:

```js
var foo = [1, 2, 3, 4]
(Array.isArray(foo) ? foo : [foo])
  .filter(Boolean)
  .forEach(function(arg){
    console.log(arg)
  })

**Comment by mattdesl** [comment]
Another way to tackle this is stop thinking of `"semi"` as an "on" or "off" type of thing. Instead; it should be: *Are you using ASI or not?*

If you aren't using ASI, lines must end with semicolons.
If you *are* using ASI, lines must not end with semicolons, but lines starting with `[`, `(`, `-`, must have a preceding semicolon.

But styles like [feross/standard](https://github.com/feross/standard) disallow preceding semicolons altogether, which is why eventually people will want these to be two independent rules ("semi" and "preceding-semi" or something).

**Comment by nzakas** [comment]
@mattdesl keep in mind you can always create your own rules to do whatever you want. The issue at hand is whether a change is needed in ESLint itself and if so, what is that change. Regardless what we decide, there's nothing stopping you from creating a rule that does exactly what you want.

> If you aren't using ASI, lines must end with semicolons.
If you are using ASI, lines must not end with semicolons, but lines starting with [, (, -, etc. must have a preceding semicolon.

All rules in ESLint core need to be standalone and relevant regardless of the other rules being used, so while it's nice to say "if semi is never, this rule is useful", that creates an implicit dependency that is both confusing and hard to explain. That means our two options are 1) to put this behavior directly into `semi` or 2) create a new rule that makes sense regardless of what `semi` is doing.

**Comment by feross** [comment]
> But styles like feross/standard disallow preceding semicolons altogether

Not exactly. It's discouraged, since these types of statements can often be rewritten to be clearer. But if you must:

```js
;[1, 2, 3].forEach(function (num) {
  console.log(num)
})
```

is not considered an error.

**Comment by glenjamin** [comment]
It seems like this rule might be useful to have on by default actually, the following snippet does not trigger a warning from the `semi` rule:

```js
var x = y
[1].forEach(console.log);
```

**Comment by rlidwka** [comment]
It should take into account indentation.

For example, this is a perfectly valid code:

```js
var str = 'foo'
        + 'bar'
        + 'baz'
```

And this is a mistake:

```js
var str = '0'
+str || console.log('str is zero')
```

**Comment by glenjamin** [comment]
As implemented at the moment I only did `[` and `(`.

Dealing with + and - is much harder, the rule which prohibits bare expressions is probably a better bet there.

**Comment by rlidwka** [comment]
Doesn't matter, this could be also a valid code:

```js
var a_long_variable_name = another_very_long_variable_name
                             ['some_long_property'].filter(Boolean)
```

My point is: we should detect developer's intent based on indentation.

**Comment by jedwards1211** [comment]
Actually, I'd opt for a rule banning `[`, `(`, `++`, and `--` at the beginning of a *statement* rather than a line; I try to avoid writing such statements anyway.

**Comment by jrajav** [comment]
@jedwards1211 That would entirely disallow the semicolon-first style, though, since the semicolon begins a new statement.

**Comment by platinumazure** [comment]
This discussion has been closed for a while. If there is a new discussion that needs to take place, could we create a new issue please? Thanks!
